### DISCUSSION TITLE
Converting image to unstructured grid in real object dimension

### DISCUSSION THREAD

**Discussion by venugovh** [pyvista/pyvista#1484] [Q&A] [answered]
I have to perform image processing on slice images of a voxel model (for the slices, I am using capturing screenshots in XY view and saving them as NumPy arrays) and convert those post-processed slices back to the voxelized model. The voxelized model thus created needs to be of the same size as the original voxel model.

The process is:

STL import >> Voxelize >> Slicing along Z >> save screen shot of each slice  as 2D array >> perform image processing to form binary 2D array

I am able to use a binary 2D NumPy array and construct an Unstructured grid that resembles the voxel model of that particular image.

The issue I am encoutering is that this voxel model is not the same size as the original voxel model since the slice images are magnified. How do I resize the voxel model to match the voxelized slice dimensions?

Couple of ways I thought this could be done. Please let me know if this is wrong!
1. I think, I could scale the voxelized model 'if I knew the pixel spacing' from the screenshot images, but I do not.
2. Slice a voxelized model and save it as a 2D matrix of 1 and 0s, basically, where ever there is material. Then I could just reconstruct the slices and translate them to the appropriate location based on the slice.center property. Is this possible?

I have attached two images to make things clear. The blue is the voxel model and red is the Unstructured grid (only some slices) with mismatched dimensions.

Could you please help me out!

![image1](https://user-images.githubusercontent.com/47824150/124369199-24952c00-dc37-11eb-9142-df3244eac298.PNG)
![image2](https://user-images.githubusercontent.com/47824150/124369200-24952c00-dc37-11eb-9754-5d6d99adfe80.PNG)

**Comment by akaszynski** [comment, marked as answer]
> save screen shot of each slice as 2D array >> perform image processing to form binary 2D array

This sounds like an unneeded step.  I think you should be able to form a binary 2D array directly from the slice.

**Reply by venugovh** [reply]
Thank you so much for your reply! I am fairly new to pyvista! _**Could you please tell me how I would go about doing that?**_ 

I think if I can get a slice output for every layer (as a binary matrix) relative to the bounding box frame, then, I don't need to worry about scaling the Unstructured Grid to match the original voxel. All the processing will be done in one scale of reference.

Thanks!

**Reply by akaszynski** [reply]
If possible, can you upload a minimum working example of your problem here?

**Reply by venugovh** [reply]
Sure, I will try to upload something as soon as possible. It is a rather long code, so will have to shorten it to be more readable.

Essentially, if I can convert the voxel data into slices of 1/0 numpy arrays (that is where I am getting stuck), then I should be able to reconstruct it back to an Unstructured grid using the original voxel size to generate the original voxelized geometry. 

I will definitely upload an example for your reference. Thanks a lot for your reply.

**Reply by venugovh** [reply]
I was able to generate a voxelized version using binary data. Another discussion on a similar note came in handy! I appreciate your inputs to this. Thank you!

https://github.com/pyvista/pyvista-support/issues/141

**Reply by akaszynski** [reply]
Thanks for posting the solution that worked!
